syntax = "proto3";

// Message to hold information for a single test case.
// You can use a C-driver to test student code without stdin/stdout.
message TestCase {
  string name = 1;
  string stdin = 2;
  string expected_out = 3;
  uint32 timeout_ms = 4;
  repeated string cl_args = 5;
  bool hidden = 6;
  uint32 points = 7;
}

// Message with information on how to measure the source code documentation.
// commands is a list of commands to generate one double (x) based on source;
// x is a percentage and corresponds to the documentation grade out of 100.
message MeasureSourceDocumentation {
  repeated string commands = 1;
}

// Message with information on how to measure the source code efficiency.
// commands should calculate one int (x) from analyzing the source file;
// points is a grade distribution, where key is a cutoff and val the percentage;
//  - e.g., {20: 1, 22: 0.95, ...}, where x <= 20 is 100%, 20<x<=22 is 95%, etc;
// default_points is pts assigned to a value of x that didn't meet any cutoff.
message MeasureSourceEfficiency {
  repeated string commands = 1;
  map<uint32, double> points = 2;
  optional double default_points = 3;
}

// Message with information on how to measure the executable efficiency.
// commands should calculate one int (x) from an accumulated execution count;
// points is a grade distribution, where key is a cutoff and val the percentage;
//  - e.g., {20: 1, 22: 0.95, ...}, where x <= 20 is 100%, 20<x<=22 is 95%, etc;
// default_points is pts assigned to a value of x that didn't meet any cutoff.
message MeasureExecEfficiency {
  repeated string commands = 1;
  map<uint32, double> points = 2;
  optional double default_points = 3;
}

// Message to hold all the information regarding the project.
message ProjectConfig {
  string name = 1;
  string rootfs_arch = 2;
  string user_filename = 3;
  string exec_name = 4;
  repeated string as_flags = 5;
  repeated string ld_flags = 6;
  repeated TestCase tests = 7;
  bool stop_on_first_test_fail = 8;
  bool must_pass_all_tests = 9;
  optional MeasureSourceDocumentation docs = 10;
  optional MeasureSourceEfficiency source_eff = 11;
  optional MeasureExecEfficiency exec_eff = 12;
  map<string, uint32> weights = 13;
  map<string, string> extra_text_files = 14;
  map<string, bytes> extra_bin_files = 15;
}

message WrappedProject {
  ProjectConfig config = 1;
  bytes checksum = 2;
}
